_source/_posts/2022-04-21-jhipster-k8s-azure.md

```diff
@@ -0,0 +1,669 @@
+
+The `ENCRYPT_KEY` will be used as the key for encrypting sensitive values stored in the Spring Cloud Config and used by the JHipster registry. You can put whatever value you want in there. A UUID works well, but any string value will work. The longer, the better.
+
+`docker-compose/.env`
+
+```.env
+SPRING_DATA_MONGO_URI="<your-connection-string>"
+ENCRYPT_KEY=<your-encryption-key
+```
+
+#### Configure Okta OAuth
+
+Use the Okta CLI to create an OIDC (OpenID Connect) application. This is what you need on the Okta side to use Okta as an authentication provider.
+
+If you already have an Okta account, use `okta login` to log into that account with the CLI. Otherwise, use `okta register` to sign up for a free account. 
+
+Create the OIDC app using the following command using a Bash shell opened to the project root.
+
+```bash
+okta apps create jhipster
+```
+
+You can accept the default values by pressing **enter**. If you want to give the app a different, more descriptive, name, that is fine. Just take note of the name because you will need to find the app in the Okta developer dashboard and update the redirect URIs a little later.
+
+This command created a config file named `.okta.env`. It will look something like the following. It helpfully lists the values you will need in the next step.
+
+```bas
+export SPRING_SECURITY_OAUTH2_CLIENT_PROVIDER_OIDC_ISSUER_URI="https://dev-123456/oauth2/default"
+export SPRING_SECURITY_OAUTH2_CLIENT_REGISTRATION_OIDC_CLIENT_ID="2989u928u383..."
+export SPRING_SECURITY_OAUTH2_CLIENT_REGISTRATION_OIDC_CLIENT_SECRET="09328uu098u4..."
+```
+
+Both the Docker Compose project and the Kubernetes project use Spring Cloud Config to centralize configuration. The JHipster registry service makes these config values available to all of the other services in the cluster.
+
+Open `docker-compose/central-server-config/application.yml` and add the following to the end, filling in the `issuer-uri`, `client-id`, and `client-secret` taken from the `.okta.env` file. This is the Spring Cloud Config file for the Docker Compose project.
+
+`docker-compose/central-server-config/application.yml`
+
+```yml
+spring:
+  security:
+    oauth2:
+      client:
+        provider:
+          oidc:
+            issuer-uri: https://<your-okta-domain>/oauth2/default
+        registration:
+          oidc:
+            client-id: <client-id>
+            client-secret: <client-secret>
+```
+
+#### Build the Docker Images and Run the App with Docker Compose
+
+You're all set to run the app locally using Docker and Docker Compose. You need to build the docker image for each of the projects: `gateway`, `store`, and `blog`.
+
+ In the three different app directories, run the following Gradle command.
+
+```
+./gradlew -Pprod bootJar jibDockerBuild
+```
```

Comment:
I get this in all three apps:

```
Downloading https://services.gradle.org/distributions/gradle-6.8.3-bin.zip
......................................................................................................
Starting a Gradle Daemon (subsequent builds will be faster)

FAILURE: Build failed with an exception.

* Where:
Settings file '/Users/mraible/Downloads/azure-k8s-cosmosdb/blog/settings.gradle'

* What went wrong:
Could not compile settings file '/Users/mraible/Downloads/azure-k8s-cosmosdb/blog/settings.gradle'.
> startup failed:
  General error during semantic analysis: Unsupported class file major version 61
  
  java.lang.IllegalArgumentException: Unsupported class file major version 61
        at groovyjarjarasm.asm.ClassReader.<init>(ClassReader.java:196)
```

```
$ java --version
openjdk 17.0.3 2022-04-19
OpenJDK Runtime Environment GraalVM CE 22.1.0 (build 17.0.3+7-jvmci-22.1-b06)
OpenJDK 64-Bit Server VM GraalVM CE 22.1.0 (build 17.0.3+7-jvmci-22.1-b06, mixed mode, sharing)
```

If I use Java 11, it works.

Reply:
Hmm. Do I think this is copied from the k8s directory of the tutorial you did a while back. Do I need to update it? Or should we just specify Java 11?

Reply:
It looks like Java 11 needs to be used. From `build.gradle` in each app:

```
sourceCompatibility=11
targetCompatibility=11
```

Reply:
OK. I'll add a note and a link to SDKMAN as a suggested solution for managing java versions.